src/FakeItEasy/Creation/DummyCreationSession.cs

```diff
@@ -0,0 +1,26 @@
+﻿namespace FakeItEasy.Creation
+{
+    using System;
+    using System.Collections.Generic;
+
+    internal class DummyCreationSession
+    {
+        private readonly HashSet<Type> typesCurrentlyBeingResolved = new HashSet<Type>();
+
+        public bool TryBeginToResolveType(Type typeOfDummy)
+        {
+            if (this.typesCurrentlyBeingResolved.Contains(typeOfDummy))
+            {
+                return false;
+            }
+
+            this.typesCurrentlyBeingResolved.Add(typeOfDummy);
+            return true;
+        }
+
+        public void OnSuccessfulResolve(Type typeOfDummy)
+        {
+            this.typesCurrentlyBeingResolved.Remove(typeOfDummy);
+        }
+    }
+}

```

Comment:
This makes sense. The extension point discovery is done only once, so there's no reason to change strategies later.